File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/ProxyBinaryMessenger.kt
package com.mapbox.maps.mapbox_maps

import io.flutter.plugin.common.BinaryMessenger
import java.nio.ByteBuffer

// BinaryMessenger that proxies all calls to another BinaryMessenger,
// but with a suffix appended to the channel name.
// Different map instances use different suffixes to avoid channel name conflicts.
class ProxyBinaryMessenger(
  private val messenger: BinaryMessenger,
  private val nameSuffix: String,
) : BinaryMessenger {

  private fun String.appendSuffix() = "$this$nameSuffix"

  override fun send(channel: String, message: ByteBuffer?) {
    messenger.send(channel.appendSuffix(), message)
  }

  override fun send(channel: String, message: ByteBuffer?, callback: BinaryMessenger.BinaryReply?) {
    messenger.send(channel.appendSuffix(), message, callback)
  }

  override fun setMessageHandler(channel: String, handler: BinaryMessenger.BinaryMessageHandler?) {
    messenger.setMessageHandler(channel.appendSuffix(), handler)
  }
}

File: lib/src/proxy_binary_messenger.dart
import 'dart:ui' as ui;
import 'package:flutter/services.dart';

class ProxyBinaryMessenger implements BinaryMessenger {
  ProxyBinaryMessenger(
      {required String suffix, BinaryMessenger? binaryMessenger})
      : _suffix = suffix,
        _binaryMessenger =
            binaryMessenger ?? ServicesBinding.instance.defaultBinaryMessenger;

  final BinaryMessenger _binaryMessenger;
  final String _suffix;

  @override
  Future<void> handlePlatformMessage(
    String channel,
    ByteData? data,
    ui.PlatformMessageResponseCallback? callback,
  ) {
    return _binaryMessenger.handlePlatformMessage(
        "$channel$_suffix", data, callback);
  }

  @override
  Future<ByteData?>? send(String channel, ByteData? data) {
    return _binaryMessenger.send("$channel$_suffix", data);
  }

  @override
  void setMessageHandler(
    String channel,
    MessageHandler? handler,
  ) {
    _binaryMessenger.setMessageHandler("$channel$_suffix", handler);
  }
}
